src/runtime/d3d12compute.cpp

```diff
@@ -86,40 +86,52 @@ class StackPrinter : public Printer<type, length> {
 static void *const user_context = NULL;
 //
 #if HALIDE_D3D12_TRACE
-// it's better to use StackPrinter<> instead of Printer<> (debug(ctx)) when
-// tracing calls because Printer<> calls 'halide_malloc()' which can fail;
-// there's even a test for it (generator_aot_cleanup_on_error), simulating
-// a halide_malloc failure, and user code should not affect debug-tracing
-#define trace(x) StackPrinter<4096, BasicPrinter>(x)
-
-static volatile ScopedSpinLock::AtomicFlag indent_lock = 0;
-static const char indent_pattern[] = "   ";
-static char indent[2048] = {};
-static int indent_end = 0;
-#define TRACEINDENT ((const char *)indent)
-#define TRACEPRINT(msg) trace(user_context) << TRACEINDENT << msg;
+static volatile ScopedSpinLock::AtomicFlag trace_indent_lock = 0;
+#define TRACE_BUF_SIZE 4096
+static const char trace_indent_pattern[] = "   ";
+static char trace_indent[TRACE_BUF_SIZE] = {};
+static char trace_buf[TRACE_BUF_SIZE] = {};
+static int trace_indent_end = 0;
+#define TRACEINDENT ((const char *)trace_indent)
+#define TRACEPRINT(msg)                                                                       \
+    {                                                                                         \
+        trace_scope___.lock();                                                                \
+        Printer<BasicPrinter, TRACE_BUF_SIZE>(user_context, trace_buf) << TRACEINDENT << msg; \
+        trace_scope___.unlock();                                                              \
+    }
 struct TraceLogScope {
-    TraceLogScope() {
-        while (__atomic_test_and_set(&indent_lock, __ATOMIC_ACQUIRE)) {
+    void *user_context;
+
+    void lock() {
```

Comment:
nit: it's too bad we can't reuse the guts of ScopedSpinLock here somehow, rather than recaptitulating the __atomic calls

Reply:
Good point. Now uses a ScopedSpinLock.